### PR TITLE
Rewrite section 3 to allow flags without response

### PR DESCRIPTION
**Pull request by yoavnir**

### CODE REVIEW

File: draft-ietf-tls-tlsflags.xml [hunk 1]
```diff
@@ -129,28 +130,27 @@
...
+            <t> Any TLS implementation that intends to propose or indicate support for a flag extension SHALL send this
+                extension with the relevant bits set.  It MUST NOT send this extension empty -- with a length of zero.</t>
+            <t> This specification does not require every flag extension to be acknowledged. Acknowledging a flag extension
+                is usually needed to inform the peer proposing the extension that the other side understands and supports
+                the extension, but some extensions do not require this acknowledgement. For example, the extension in
+                <xref target="I-D.kampanakis-tls-scas-latest" /> allows the peer to send an abbreviated certificate chain.
+                The peer need not acknowledge that it supports sending an abbreviated chain. It's enough that it sends 
+                the chain abbreviated. It is up to the document defining a particular flag extension to specify whether that
+                extension requires an acknowledgement or not.</t>
+            <t> Assuming that an acknowledgement is required, it SHOULD come at the earliest possible opportunity. A
```
Comment (martinthomson) [inline draft-ietf-tls-tlsflags.xml:142]:
```suggestion
            <t>A
```

Earliest possible is probably not right.

1. SH is earlier than EE, but EE is usually the better choice.
2. CT is later than EE, but if the extension relates to the handling of a certificate, that is the right place.
3. HRR is earlier than anything, but unless the flag is necessary for the client to generate a new CH, HRR is the wrong place.

File: draft-ietf-tls-tlsflags.xml [hunk 2]
```diff
@@ -129,28 +130,27 @@
...
+                flag proposed by the client in ClientHello (CH) SHOULD be acknowledged in either ServerHello (SH), in 
+                EncryptedExtensions (EE), in Certificate (CT), or in HelloRetryRequest (HRR) as the flag document specifies. 
+                A flag proposed by the server in the CertificateRequest (CR) message SHOULD be acknowledged in the client's
+                Certificate (CT) message. A flag proposed by the server in the NewSessionTicket (NST) message is never 
+                acknowledged as there is not client-side response message.</t>
+            <t> Multiple flags can be proposed or acknowledged in the same extension.</t>
+            <t> In all of the above cases, no flag is ever acknowledged in SH, EE, CT, or HRR without first having been
```
Comment (martinthomson) [inline draft-ietf-tls-tlsflags.xml:149]:
```suggestion
            <t> In all of the above cases, a flag MUST NOT be acknowledged in SH, EE, CT, or HRR without first having been
```

File: draft-ietf-tls-tlsflags.xml [hunk 3]
```diff
@@ -129,28 +130,27 @@
...
+                extension requires an acknowledgement or not.</t>
+            <t> Assuming that an acknowledgement is required, it SHOULD come at the earliest possible opportunity. A
+                flag proposed by the client in ClientHello (CH) SHOULD be acknowledged in either ServerHello (SH), in 
+                EncryptedExtensions (EE), in Certificate (CT), or in HelloRetryRequest (HRR) as the flag document specifies. 
+                A flag proposed by the server in the CertificateRequest (CR) message SHOULD be acknowledged in the client's
+                Certificate (CT) message. A flag proposed by the server in the NewSessionTicket (NST) message is never 
+                acknowledged as there is not client-side response message.</t>
+            <t> Multiple flags can be proposed or acknowledged in the same extension.</t>
+            <t> In all of the above cases, no flag is ever acknowledged in SH, EE, CT, or HRR without first having been 
+                proposed in CH or CR. Unsolicited flags may appear only in CH, CR, and NST.</t>
```
Comment (martinthomson) [inline draft-ietf-tls-tlsflags.xml:150]:
```suggestion
                proposed in CH or CR. Unsolicited flags may appear only in CH, CR, and NST.
                And endpoint that receives an unsolicited flag in another message (HRR, SH, EE, or CT)
                MUST generate a fatal illegal_parameter alert.</t>
```

**Review by martinthomson — COMMENTED** [review]
I'm not sure that you want to publish a document with a reference (even if informative) to the scas draft.  Other than that, this is the right way to structure this, provided a couple of issues are resolved.  I've suggestions for each.

**Comment by richsalz** [comment]
Does it make sense to define a "echo this back" flag?  Just thinking out loud.  Argument against it is that clients will just always add that flag/extension ...

**Comment by martinthomson** [comment]
An empty extension remains as a cheaper option.

File: draft-ietf-tls-tlsflags.xml [hunk 4]
```diff
@@ -129,28 +130,27 @@
...
         <section anchor="rules" title="Rules for The Flags Extension">
+            <t> Any TLS implementation that intends to propose or indicate support for a flag extension SHALL send this
+                extension with the relevant bits set.  It MUST NOT send this extension empty -- with a length of zero.</t>
+            <t> This specification does not require every flag extension to be acknowledged. Acknowledging a flag extension
+                is usually needed to inform the peer proposing the extension that the other side understands and supports
+                the extension, but some extensions do not require this acknowledgement. For example, the extension in
+                <xref target="I-D.kampanakis-tls-scas-latest" /> allows the peer to send an abbreviated certificate chain.
+                The peer need not acknowledge that it supports sending an abbreviated chain. It's enough that it sends 
+                the chain abbreviated. It is up to the document defining a particular flag extension to specify whether that
+                extension requires an acknowledgement or not.</t>
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-ietf-tls-tlsflags.xml:141]: I don't think the example here is necessary.

File: draft-ietf-tls-tlsflags.xml [hunk 5]
```diff
@@ -129,28 +130,27 @@
...
+            <t> Assuming that an acknowledgement is required, it SHOULD come at the earliest possible opportunity. A
+                flag proposed by the client in ClientHello (CH) SHOULD be acknowledged in either ServerHello (SH), in 
+                EncryptedExtensions (EE), in Certificate (CT), or in HelloRetryRequest (HRR) as the flag document specifies.
```
Comment (chris-wood) [inline draft-ietf-tls-tlsflags.xml:144]:
```suggestion
                EncryptedExtensions (EE), in Certificate (CT), or in HelloRetryRequest (HRR) as the corresponding flag document specifies. 
```

File: draft-ietf-tls-tlsflags.xml [hunk 6]
```diff
@@ -129,28 +130,27 @@
...
+                extension with the relevant bits set.  It MUST NOT send this extension empty -- with a length of zero.</t>
+            <t> This specification does not require every flag extension to be acknowledged. Acknowledging a flag extension
+                is usually needed to inform the peer proposing the extension that the other side understands and supports
+                the extension, but some extensions do not require this acknowledgement. For example, the extension in
+                <xref target="I-D.kampanakis-tls-scas-latest" /> allows the peer to send an abbreviated certificate chain.
+                The peer need not acknowledge that it supports sending an abbreviated chain. It's enough that it sends 
+                the chain abbreviated. It is up to the document defining a particular flag extension to specify whether that
+                extension requires an acknowledgement or not.</t>
+            <t> Assuming that an acknowledgement is required, it SHOULD come at the earliest possible opportunity. A
+                flag proposed by the client in ClientHello (CH) SHOULD be acknowledged in either ServerHello (SH), in
```
Comment (chris-wood) [inline draft-ietf-tls-tlsflags.xml:143]:
```suggestion
                flag proposed by the client in ClientHello (CH) that requires acknowledgement SHOULD be acknowledged in either ServerHello (SH), in 
```

File: draft-ietf-tls-tlsflags.xml [hunk 7]
```diff
@@ -129,28 +130,27 @@
...
+                is usually needed to inform the peer proposing the extension that the other side understands and supports
+                the extension, but some extensions do not require this acknowledgement. For example, the extension in
+                <xref target="I-D.kampanakis-tls-scas-latest" /> allows the peer to send an abbreviated certificate chain.
+                The peer need not acknowledge that it supports sending an abbreviated chain. It's enough that it sends 
+                the chain abbreviated. It is up to the document defining a particular flag extension to specify whether that
+                extension requires an acknowledgement or not.</t>
+            <t> Assuming that an acknowledgement is required, it SHOULD come at the earliest possible opportunity. A
+                flag proposed by the client in ClientHello (CH) SHOULD be acknowledged in either ServerHello (SH), in 
+                EncryptedExtensions (EE), in Certificate (CT), or in HelloRetryRequest (HRR) as the flag document specifies. 
+                A flag proposed by the server in the CertificateRequest (CR) message SHOULD be acknowledged in the client's
```
Comment (chris-wood) [inline draft-ietf-tls-tlsflags.xml:145]:
```suggestion
                Similarly, a flag proposed by the server in the CertificateRequest (CR) message that requires acknowledgement SHOULD be acknowledged in the client's
```

File: draft-ietf-tls-tlsflags.xml [hunk 8]
```diff
@@ -231,6 +231,7 @@
             &I-D.sy-tls-resumption-group;
             &I-D.ietf-tls-tls13-cert-with-extern-psk;
             &I-D.ietf-tls-cross-sni-resumption;
+            &I-D.kampanakis-tls-scas-latest;
```
Comment (chris-wood) [inline draft-ietf-tls-tlsflags.xml:234]:
```suggestion
```

**Pull request merged**